### ISSUE TITLE
3U DDS (URUKUL) discussion

### ISSUE THREAD

**Issue by gkasprow** [m-labs/sinara#191]
I did early estimation of available board area.
It seems that we can easily fit 4 channels on 100x160 Eurocard.
There are several questions which need to be answered before we continue with schematics:
- I assume we don't need clock shaping circuit in AD9912 and we will route DAC output (pins 50,51) directly to the circuit copied from Allaki.
- do we need to measure output level as it is done in Allaki design? Should I remove AD8363 or add some I2C/SPI ADC ?
- we need to control several IO lines per channel. Which one can be common and which can be common?
I assume that DDS Reset, S4...S1, IO_update, CLKMODESEL can be common while PWRDOWN should be individual. For output channel switch, we would need individual RFSW control. Attenuator can have common LE, CLK and RSTn.
- How we distribute SYSCLK? I assume we have SMA input on front panel and some other connector (i.e. mmcx) on the other side of PCB to route via jumpers to Kasli clock distribution. I assume we will use 1:4 clock splitter and input circuit as it is on Sayma RTM.
- I copy loop filter from devkit
- What about SPI control? I can connect 4x HMC542 to have 32bit data and one CS. I can also connect them as 2x 16bit but additional CS will be needed.
- the same I can do with SPI registers (SN74LV595) to have 32it data.
- I will use 8 output address decoder ('138) to connect to all CSn lines. We can route additional CSn or treat state 000 as none SPI selected. SPI address decoder map:
000 - no slave selected
001 - DDS1
010 - DDS2
011 - DDS3
100 - DDS4
101 - attenuators CS
110 - serial register CS
111 - reserved

- LVDS assignment: 
LVDS1: SCKI
LVDS2: SDI
LVDS3: SDO
LVDS4: SEL CSN3
LVDS5: SEL CSN2
LVDS6: SEL CSN1
LVDS7: IO_UPDATE ??
LVDS8: RESET ??

- for supply I want to use similar double output converter modules as for Sayma RTM + LDOs regulating down to +5V and 3.3/1.8V
- do we plan to compensate for clock delays? We can use one of clock outputs and route it back to Kasli phase detector.
- below is initial component placement on PCB. Each channel will have individual shield.

![obraz](https://cloud.githubusercontent.com/assets/4325054/26056705/c54f1178-3976-11e7-8d4b-84065c1de8b6.png)

**Comment by hartytp** [comment]
@gkasprow Looks nice!

@jordens How do you want to manage the review process for Urukul+ Kasli? Do you want feedback from us at this stage? I can guess the answers to most of these questions, but I won't add my tuppence worth unless it's wanted!

**Comment by hartytp** [comment]
Can you remind me what the plan for clock distribution is? Is the idea that we'll distribute a ~100MHz clock to Kasli + Urukul (via SMAs/backplane/CDR from DRTIO) and then use local PLLs to generate the DAC clock? Will you use the PLL inside the DAC, or add an external IC (HMC830 or similar)?

**Comment by gkasprow** [comment]
At the moment the clock can be taken either from front panel SMA or Kasli clock distribution. I assume we use the PLL in DDS chip.

**Comment by dhslichter** [comment]
While we are in the early stages: we might consider using AD9914 instead of AD9912, for two reasons:

- profile mode
- potential for higher clock frequency (thus higher output frequencies).

Particularly for DDS chips which are programmed over serial, profile mode is very handy because it can allow fast frequency/phase/amplitude hopping with deterministic timing at faster timescales than one can reprogram everything over SPI.  We use this methodology for our microwave generation in the magtrap.  The user does not need to use profile mode if they don't need/want it, but it can be handy e.g. for hopping around the hyperfine manifold.  

Also, a 1 GHz clock on the DDS limits us to ~400 MHz (assuming we put reconstruction filters etc on the Urukul board).  If you go with the AD9914, you can run at e.g. 2 GHz clock frequency, which lets you go up to ~800 MHz output frequencies, which is handy for some groups that drive 600 MHz AOMs, for example.  We run ours at 2.4 GHz.  

I would use an external PLL for the reference clock generation (e.g. HMC830, something already in Sayma so the gateware will be written) because I think they tend to be better/more flexible than the internal DDS PLLs.

**Comment by sbourdeauducq** [comment]
Third reason: we already know the main silicon bugs of the 9914, but not those of the 9912.

**Comment by sbourdeauducq** [comment]
That being said, if someone from ADI can confirm that the 9912 is bug-free, I would prefer it.

**Comment by gkasprow** [comment]
Take into account that ad9914 cost is 188$ while AD9912 is 61$. This is the most expensive part on the board. It consumes 3W of power while AD9914 consumes about 1.3W. 12W (4 channels) of power dissipation just for DDS chips is quite a lot. I have devkit with 9912 so can test some functionalities.

**Comment by dhslichter** [comment]
This cost difference is not zero, but given the additional functionality I'd argue it's definitely worth it.  $120 per channel is a very small cost in the types of setups these boards would be used in.  The power dissipation is definitely an issue to be considered; however, with forced air this can certainly be handled (and this is in fact what we do with our hardware).  

The decision that needs to be made here is basically about how capable we want/need Urukul to be.  Is the target market just to do simple CW signal generation for AOMs?  Or are people going to use these to generate signals for pulsed setups (with some modulation or switch afterward)?  What kind of flexibility would be desired?  

From my standpoint, the AD9914 gives the option for substantially broader use cases, at a pretty low cost ($500/board extra, assuming a 4-channel board), and with no additional complexity on the gateware/programming side.  

One of the main issues I see is that people may want a solution which can do pulsed operations, or fast frequency/phase switching, but may not be ready to commit to the full Sayma system in the crate (which is very large and expensive compared with the Kasli/EEM ecosystem).  By changing the Urukul to the AD9914, one would be able to bridge this functionality gap more readily, and provide a more "mid-grade" performance solution.

**Comment by gkasprow** [comment]
For me the only difference at the moment is that I already purchased devkit with 9912. To do quick tests with 9914 I would need another devkit for 700$. They have different interfaces so cannot connect my board with 9914 to 9912 devkit.

**Comment by hartytp** [comment]
> @dhslichter Particularly for DDS chips which are programmed over serial, profile mode is very handy because it can allow fast frequency/phase/amplitude hopping with deterministic timing at faster timescales than one can reprogram everything over SPI. We use this methodology for our microwave generation in the magtrap. The user does not need to use profile mode if they don't need/want it, but it can be handy e.g. for hopping around the hyperfine manifold.

How do you want to connect the profile pins to Kasli? Bearing in mind that the IDC only has 8 LVDS pairs to control all 4 DDS channels, we'd need to use multiplexed IO. Depending on how we do that, it may not end up being that much faster than reprogramming the DDS over SPI.

> 12W (4 channels) of power dissipation just for DDS chips is quite a lot.

Currently, we're budgeting for 6W (0.5A of 12V) per EEM. If we go for the AD9914, we should probably increase that to more like 1.5A per EEM. That would mean increasing the current rating of the power connectors on Kasli/VHDCI carrier (currently 5A IIRC).

**Comment by dtcallcock** [comment]
How about giving users the option of using two IDCs?  In that case the timing-critical profile selects and RF switches would have dedicated lines.  It would also mean that you only have to draw 0.75A per IDC.

Even in the AD9912 case, it's not clear how the RF switches are controlled.

**Comment by jordens** [comment]
@gkasprow 
Yes. Output circuitry is allaki directly. No output power sensing. No ad8363 etc. AA filtering (and Nyquist zone selection) to be done with the allaki filters. DDS reset, io update are common, s1-4 and clkmodesel are common and can be resistor jumpers. Pwrdown can be individual resistor jumpers. RF switches should be an individual eem line (maybe an addressable latch with the DDS cs?). Attenuators can be daisy chained with a single cs. 
More later...

**Comment by hartytp** [comment]
@jordens +1 to all of that.

Also FWIW, we'd much prefer the AD9912 to the AD9914. We would also really like this design to run from a single IDC.

**Comment by hartytp** [comment]
@gkasprow When you produce the prototypes for this board please can we buy 5? We'd like these asap, and don't mind using prototype hardware...

**Comment by gkasprow** [comment]
@hartytp So we will make at least 6 prototypes of DDS? One stays at WUT, other go abroad. Anybody else wants first prototypes?

**Comment by hartytp** [comment]
@gkasprow Thanks! That sounds good. Just send me the quote when you're ready.

**Comment by dhslichter** [comment]
I agree with @dtcallcock; my vision was that you have two IDCs to cover the 4 channels (allowing for profile select, plus fast rf switch open/close).

**Comment by dhslichter** [comment]
I doubt NIST will be buying any of these boards if the AD9912 is used.  That is not intended as an attempt to influence the design, but rather just informational -- it would just be a major step below what we already have in terms of capability.  I also think that one could end up making two board designs: an Urukul "basic DDS" board and another board with higher performance (fast switching, higher clock frequency) featuring AD9914s.  This of course doubles the design and prototyping work and requires additional funding from someone else.  I do think that the market for fast switching and higher output frequencies is substantial, and that it should be an important consideration for developing an ARTIQ hardware ecosystem in general.  I recognize that Oxford has their own set of priorities here, and that they are focused on that.

**Comment by gkasprow** [comment]
@dhslichter We can always make two wersions of DDS board. One - simplified with minimal cost, second - high performance used with two IDCs.

**Comment by dhslichter** [comment]
@hartytp just curious, why the "much prefer" for AD9912?

**Comment by dhslichter** [comment]
@gkasprow you read my mind :)

**Comment by gkasprow** [comment]
@hartytp @dhslichter @jordens Guys, we can make variant design.
We have some spare area and can place both DDS chips on both side of the board.
When you order board you simply specify the variant 
So we have same PCB and 2 versions of pick&place files which result in different board assembly.
So ones who want 9912 will get it mounted with 1 IDC, others who want 9914 will get the same board but with other chip and two IDCs.
This adds some work  so additional founding would be very welcome. I'd like to buy 9914 devkit to fully characterise the board.

**Comment by sbourdeauducq** [comment]
This sounds quite complicated. Can we please keep the Kasli ecosystem simple?

**Comment by gkasprow** [comment]
@sbourdeauducq This is low cost alternative to having two boards design. We can even call them differently :)

**Comment by hartytp** [comment]
@dhslichter Maybe "_much_ prefer" is too strong, but prefer nonetheless. To explain our priorities:

We are setting up an experiment which will have two ion traps, each with Ca + Sr. We will use Sayma for qubit manipulations, which require pulse shaping etc. For everything else (cooling AOMs etc) we hope to use Urukul, with Novogorny for noise eating (initially in software, but eventually in gateware).

Given that we need 60+ channels of RF for this, both cost and power consumption are important to us (our labs' AC is only 10kW). As the AD9912 can do everything we would want, and is cheaper and less power hungry, we prefer it.

Beyond that, I think there's a lot of advantage to keeping the Kasli ecosystem as simple and low cost as possible and avoid adding features that aren't actually needed.

IMO the profiles on the AD9914 aren't that useful since one can reprogram the DDS over SPI in ~1us, which is fast enough for most cases. Using up two IDCs for this design is a pain: more ribbon cable wiring (or backplane sockets taken up); and reduces the number of these one can run from a Kasli. This is particularly true for our initial experiments, where we want to run as many of these as possible from a single KC705.

**Comment by sbourdeauducq** [comment]
If we end up with both 9912 and 9914, I would prefer two board designs, and keep each one as simple as possible.

**Comment by gkasprow** [comment]
@sbourdeauducq Both designs differ just by one chip and having variants cuts half of development costs. So we don't loose anything. If someone wants to fund second design and prototype round, I'm open:)

**Comment by dtcallcock** [comment]
If there were to be two boards it would also allow better optimization for Tom's use case.  For example, doubled AD9911 could be used at 1/2 cost and power consumption.  

The 4ch AD9959 could also be used to go to 8ch per card for the same cost/power.  This would increase development costs though as the two boards would be quite different.

**Comment by hartytp** [comment]
> @dtcallcock If there were to be two boards it would also allow better optimization for Tom's use case. For example, doubled AD9911 could be used at 1/2 cost and power consumption.

I'd prefer to avoid frequency doubling if possible, since it's nice to be able to linearly vary the output amplitude by programming the DDS amplitude.

**Comment by hartytp** [comment]
Anyway, I don't want to take this conversation too far since we didn't fund this project and it's not our decision. I'm just stating a preference, but in the end we'd buy the boards either way.

**Comment by jordens** [comment]
@gkasprow
The RF switches need to occupy eem lines directly. Plus scki SDI sdo selX io-update. The rest can go on the shift register. With two DDS per board we could do that with just one eem. Four would need two eem and would mean more spin bus contention. I am still undecided whether two or four DDS is the best approach here.
Loop filter from devkit is good. My guess is that the 10x setting should be default. AFAICT we can have both pll (with e.g. 100 MHz to urukul) and direct drive clock (e.g. 1 GHz) with just the clockmodesel. Simple dumb fanout to the DDS is ok. With inputs from the mmcx, backplane or the front  panel. If necessary with capacitor jumpers. I don't think we need a hmc830 here.
Clock delays don't matter.
 The supply sounds good.
@hartyp the current modus operandi is good. All comments and review are highly appreciated.
@dhslichter ad9912 covers our use cases just fine. There are Nyquist  images for higher freuqencies. Profile mode is not something we need right now. And we cant afford the eem pins. For more  features we have sayma. This is very much the 'simple aom' driver with RF switch and attenuators you mentioned.
@gkasprow I am ok with allowing the ad9914 as a bom alternative. But there should not be noticable compromises. Not in the number of board layers, not In The number of eem connectors required. Not in the power supply budget or board price or RF shielding quality...
We'll have one prototype for QUARTIQ  then two in the second batch.
@dtcallcock the ad9911 and ad9959 would be great if they did 1gsps, 48 bit ftw.

**Comment by hartytp** [comment]
> The RF switches need to occupy eem lines directly.
> I am still undecided whether two or four DDS is the best approach here.

In our case, where a large number of channels are needed, 4 DDS per IDC (and ideally, 4 DDS per card) would be much nicer than 2. For example, this allows one to run a fairly complex experiment from the FMCs on a KC705 without having to buy an additional bunch of Kaslis -- this would be particularly helpful until the bugs are ironed out of DRTIO + Kasli support.

Why do the RF switches need to be connected directly to the IDC? What maximum pulse rate do you want to support? I was hoping we could get away with an SPI -> 8 TTL type IC to control them (same bus as the DDS, using CS lines).

**Comment by gkasprow** [comment]
We can use DDR SCLK and two available LVDS lines to control 4 switches – rising edge controls one switch, falling controls another.

One quad flip-flop would solve the issue.

**Comment by dhslichter** [comment]
Agreed that a shift register with latching output would be a suitable method of doing the rf switches -- you get the clean timing control with a single latch line.  But it does restrict the pulse durations to ~1 us or more.  Now, if the goal is a 'simple AOM driver', that is probably fine.

**Comment by jordens** [comment]
From our view this is less about pulse rate and more about being able to simultaneously switch and with sufficient time resolution and unencumbered by bus contention. Otherwise we could just zero the ftw and would not really  need switches at all.

**Comment by jordens** [comment]
@hartytp if you want to hook this up to a custom adapter board for kc705 you can bus-share spi and shift-register the switches there. The standardized eem IDC pinout allows it.
Would having 2 IDC for 4 dds channels really be too much for you?

**Comment by gkasprow** [comment]
The LVDS goes up to 150MHz easily, so with 8 bit shift register one can get 100ns or less IO toggle rate. Would it be too high period for switches and profile changes?

**Comment by hartytp** [comment]
> @gkasprow The LVDS goes up to 150MHz easily, so with 8 bit shift register one can get 100ns or less IO toggle rate. Would it be too high period for switches and profile changes?

Agreed. Given that most AOM rise times are at least a few hundred ns, I don't see this as being an issue.

> @jordens From our view this is less about pulse rate and more about being able to simultaneously switch and with sufficient time resolution and unencumbered by bus contention

With latching shift registers one can still switch channels simultaneously. This also doesn't limit the timing resolution, just the minimum gap between state changes.

> Would having 2 IDC for 4 dds channels really be too much for you?

It suspect it would be quite inconvenient, but not a killer. 

> if you want to hook this up to a custom adapter board for kc705 you can bus-share spi and shift-register the switches there. 

This is an option, but it'd be nice to get away with only the existing passive adapter boards.

> Otherwise we could just zero the ftw and would not really need switches at all.

So long as there aren't any noise/cross-talk/clock leakage/etc issues, this would be fine for us.

My preference would still be to use the shift register but, if you think that won't work for you, then how about we connect the switches to the second IDC, and set the pull ups/downs to leave them NO. That way basic functionality (DDS SPI, attenuator etc) can be accessed from a single IDC. Users who want the switches can optionally connect a second IDC.

**Comment by gkasprow** [comment]
@jordens @hartytp What is typical control sequence of DDS chip?
Do you need to simultaneously talk over SPI and update the switches or can setup the DDS registers once and toggle switches/profiles after that? In first case we need dedicated data and latch lines while in second case we don't need any additional wires since there is one free CS in address decoder.

**Comment by dhslichter** [comment]
Typical control sequence is for SPI programming to the DDS, followed by an IO_UPDATE pulse after potentially some delay.  For example, if you want to change three DDS chips simultaneously, you program each in turn and then give them all an IO_UPDATE simultaneously.  The rf switches might or might not be operated at the same time as an IO_UPDATE.  If you are changing frequencies often, you might want to be programming some DDS chips for the next frequency update at the same time that you'd like to change the state of the rf switches, thus the bus contention that @jordens alludes to.

**Comment by hartytp** [comment]
> The rf switches might or might not be operated at the same time as an IO_UPDATE.  

I assumed that you'd generally always do the IO_UPDATE before (after) opening (closing) the switch to avoid glitches. 

> If you are changing frequencies often, you might want to be programming some DDS chips for the next frequency update at the same time that you'd like to change the state of the rf switches, thus the bus contention that @jordens alludes to.

Assuming that this is just a basic source for cooling AOMs etc, the lengths of the pulses will generally be long compared with the ~1us it takes to reprogram the DDSs via SPI. Thus there should be plenty of time to reprogram the DDSs after the switches have changed states.

But, I do take the point that dealing with shared busses in a real time system is always more complicated for the user.

**Comment by gkasprow** [comment]
@dhslichter Let's assume that switches are controlled by shift register clocked by SCKI (which is active all the time) and data are transferred over dedicated SWITCHES_DAT line. Then SPI core sets SEL_CSN3..0 = 111 and latches configuration. In this way we can run both transfers in parallel without conflicts with SPI. The only difference is that SPI  returns to SEL_CSN3..0 = 111 instead of 000 after the transfer.
The signal assignment would look like this:

LVDS1: SCKI
LVDS2: SDI
LVDS3: SDO
LVDS4: SEL CSN3
LVDS5: SEL CSN2
LVDS6: SEL CSN1
LVDS7: IO_UPDATE
LVDS8: SWITCHES_DAT

**Comment by hartytp** [comment]
@gkasprow NB the AD9912 max SPI clock is 50MHz, which will limit the speed we can run the shift register at if they're on the same bus.

**Comment by gkasprow** [comment]
We can transfer the register data on two edges of the SPI clock:)

**Comment by dhslichter** [comment]
OK, with IO_UPDATE and SWITCHES_DAT are global updates that latch the most recent serial data into DDS output registers and rf switch configurations, respectively.  I think this works fine; as stated before, the bus arbitration could get complicated but the problem is basically being pushed on the end user.  

@hartytp @gkasprow the ARTIQ SPI gateware doesn't allow for DDR right now, so that would have to be redone, and I am not sure how rapidly one can change over between SDR and DDR to do e.g. DDS programming, switch changes, back to DDS programming...there is nontrivial time overhead in the current SPI implementation on the FPGA for these things.

**Comment by hartytp** [comment]
> @dhslichter the ARTIQ SPI gateware doesn't allow for DDR right now

Agreed, I expected that we'd just run the bus at single data rate at the maximum clock rate supported by the DDS. 8bits/50MHz=160ns, which is still plenty fast enough for all practical applications I can think of for this kind of hardware (i.e. it's fast compared with the rise time of most AOMs).

> the ARTIQ SPI gateware doesn't allow for DDR right now, so that would have to be redone, and I am not sure how rapidly one can change over between SDR and DDR to do e.g. DDS programming, switch changes, back to DDS programming...there is nontrivial time overhead in the current SPI implementation on the FPGA for these things.

I'm not sure the factor of two is worth the complexity of supporting DDR, so this is a bit of a moot point, but... I'd guess that if you did go down that route, you'd just the core in DDR mode for the DDS as well, and copy each bit twice. That way they DDS don't notice the difference...

**Comment by hartytp** [comment]
Putting our priorities in more context:

I'd like to be able to set up a Eurorack with

- 1 x Kasli
- 2 x Novogorny
- 4 x Urukul
- 2 x other EEM, such as BNC or Zotino

Combining this with 2 x RF PA gives a very cost/space/power efficient setup for driving 16 AOMs (inc intensity servo) with enough performance/functionality/flexibility for the majority of our use cases (cooling/optical pumping/etc). Note that, given that we will have up to 60 AOMs in a lab (FWIW, these are also rack-mounted on optical bread boards with fibre inputs/outputs), space efficiency is a real priority for us, which is why we're keen to stick with 4 DDS per IDC if possible!

Initially, we'd do noise eating in software using the gateware/software that's already been funded. In the longer term, we'd want a Kasli servo gateware that used Novogorny in continuous sample mode (125kSPS/channel).

We would use Sayma for qubit-type stuff only. In most cases, only 1-2 Sayma will be required for an experiment. In this case, we'd hope to do away with the cost/complexity of uTCA chassis + Metlino by using Kasli as the ARTIQ master and mounting the Sayma in Greg's 1 unit chassis.

We'd only fallback to uTCA chassis/Metlino for really complex experiments that required many SAWG channels or (to be funded/designed) a lot of fast DAC channels for shuttling.

**Comment by hartytp** [comment]
Thinking about this a bit more, I feel that putting the switches on the same SPI bus as the DDSs is a bad idea: in the long run, we'd like to use these boards in a servo loop (feeding back on an input to Novogorny). For this, the servo gateware will update the DDSs continuously at max rate. Doing that while sharing the SPI bus with the switches would be really nasty.

In a similar vein, to make this practical, we need to keep the time between DDS updates down to a few us. Assuming we don't want to run the DDS SPI bus as DDR (needs extra FFs), that will probably mean limiting us to 2 DDS max per SPI bus (note to self: check the timings for this).

**Comment by gkasprow** [comment]
@hartytp Can we control DDS IO_UPDATE together with switches from same register? Two lines should be enough to control such register with independent data rate

**Comment by gkasprow** [comment]
The solution is fairly simple - you transmit DDR data to register running i.e. with 150MHz clock on rising edge. On falling edge you transfer latch state. So one shift register (i.e. AHC595)+ 1 flip flop would do the job. From FPGA side it would be trivial state machine with IODDR.

**Comment by hartytp** [comment]
> @gkasprow Can we control DDS IO_UPDATE together with switches from same register? 

The more I think about it, the less comfortable I get using that kind of shared bus when the DDSs are operated in continuous update mode for a servo loop.

My preferred option would be 2 DDS per IDC (SCLK, SDO, SDI, SW1, SW2, IO_UPDATE, CS1, CS2). Ideally, 4 DDS (2 IDC) per board so we can get a decent number of channels in a Eurorack, but I don't feel too strongly about that.

Floating one other idea: what about adding a flip-flop to the board and running the SPI bus DDR (needs to be supported in the SPI core, but that's a pretty minor upgrade)? Updating the DDS via SPI takes ~1.5us. For 2 DDSs on the SPI bus, this limits the servo bandwidth, the time between frequency updates, etc to ~3us. We can live with that (just), but would be much more comfortable with 1.5us.

**Comment by gkasprow** [comment]
@hartytp We can use DDR mode and double data line to SPI.
In this way we can talk to 4 DDS chips at the same time. We would need just 2 flip-flops on the board to enable it. DDS readout can be done by common line. For this purpose there would be reserved address <111> that enables all CS lines of four DDS chips so you can talk to them independently.
Providing that we can connect UPDATE and switches line to same register and use 100MHz clock, we can both support 4DDS running in parallel and 20MHz toggle rate for switches and UPDATE line.
Signal assignment would be:
(SCLK, SDO1, SDO2, SDI, CS_AD0,CS_AD1,CS_AD2, REG_DAT)
CS address would be assigned in following way:
000 - no slave selected
001 - DDS1
010 - DDS2
011 - DDS3
100 - DDS4
101 - attenuators CS
110 - slow serial register CS
111 - all DDS chips enabled

**Comment by gkasprow** [comment]
@hartytp ADC and DAC designs can be easily adopted - just one flip flop can be added easily to prototypes after boards are produced.

**Comment by dhslichter** [comment]
But for the DAC design you don't win anything, because there is only one chip being controlled by SPI per board, right?

**Comment by gkasprow** [comment]
@dhslichter @hartytp  Yes, but by adding flip flop you can use same DDR SPI core for all boards including DAC and ADC.
All depends on setup and hold times. At 50MHz the DAC chip may not really like data transmitted on falling edge because it may violate the hold time. It requires 5ns of setup and 5ns of hold time. 50MHz is 20ns of clock period so we split it into 2 windows with data, 10ns each. And here we are on the edge of setup and hold time. By adding the flip flop you simply enlarge the data window twice. It is still compatible with SDR and DDR operating mode - simply ignores data sent on falling edge.
For DDS, you simply add two flip flops that split single data line into 2 chips.
So over the data line you send two streams interleaved in following way:
DAC1:15 DAC2:15 DAC1:14 DAC2:14, etc.
Standard DDR transfer usually does consecutive bits on rising and falling edges. For simplicity of interfaces, we use interleaved mode with two data lines in DDR mode.
 So to summarize - we can independently and simultaneously control 4 DACs at full speed with DDR clock.
Modification of SPI core would be simple - just add DDR output buffer and use four data registers to transfer 4 data streams to 4 DDS chips at the same time. Some control logic to steer CS could be needed to decide which DACs are updated - if only one or all at the time.
If you connect standard SPI core, only two channels will work independently. Two other channels will receive identical data so such approach is transparent and can run with standard SPI core in standard mode where each chip at the time receives data. When SPI core gets updated, four DDS chips may run in parallel with individual data. So with same hardware we get several possibilities.
And we get additional, third data stream for register control. Which runs all the time without SPI intervention. On FPGA side you just instantiate shift register, counter and IODDR that continuously pushes the data to our register. I can provide such code. The only condition is that SCK is available all the time, but I don't see any problem with it. Using this approach we can really quickly (a few dozens of 10ns) change profiles for 9914DDS 
For DAC board I designed little board with flip flops and gates so we can add it later by unsoldering resistors on ADC/DAC and using some wires to implement DDR functionality.
![obraz](https://cloud.githubusercontent.com/assets/4325054/26211049/cc7c2a8a-3bf1-11e7-9ff0-f99a93245ad0.png)

**Comment by sbourdeauducq** [comment]
AFAIK the switches need good timing precision, so they should have their own dedicated TTL lines and not controlled through some time-sharing mechanism on a bus.

**Comment by gkasprow** [comment]
They have dedicated data line and only shared clock with SPI that runs continuously.
Update rate is strictly controlled by a counter that counts SPI cloc cycles or other hardware "tick" that synchronizes all updates crate wide. It can be i.e. 500ns system timer or so.

**Comment by dhslichter** [comment]
@sbourdeauducq if we ignore bus contention issues for the moment, the timing precision can still be good because you use a dedicated TTL to determine the time at which the rf switch states are latched from the shift register.  The main issue comes from bus contention here.  The proposal from @hartytp and @gkasprow is to upgrade the gateware to run SPI DDR, at ~100-150 MHz clock speeds, and use this to multiplex multiple SPI targets on a single physics bus while running each at maximum speed.  I have to say, it seems like it will be a headache to pack and unpack the multiplexed data on the SPI master end (in the FPGA).  I think that subtleties and corner cases with chip selects or different length word writes etc will probably be a pain to manage.

**Comment by hartytp** [comment]
Separating two different issues here: (1) whether we multiplex the RF switch TTLs, or just run them directly from the IDC; and (2) how we update the DDS chips.

(1) While I agree that @gkasprow's proposal would work in principle, I also agree with @sbourdeauducq and @dhslichter that it wouldn't fit in well with the way we would use this hardware in the lab. I think the extra complexity it introduces isn't worth saving the IDC. So, I'd prefer to run the switches directly from the IDC (1 LVDS pair per switch).

(2) As I mentioned above, our main use case for this hardware would be running the DDSs in "constant update" mode as part of an intensity servo. This is a bit messy because the user can only change the DDS frequency/phase/amplitude at fixed times, but there isn't a way of getting around that without introducing an extra amplitude modulator of some form after the DDS. In practice (aside from the servo loop), the user would only update the DDSs while the switches are shut, so that they aren't sensitive to the DDS update timing.

If we do put 2 DDSs on the same SPI bus then the delay between updates is ~3us, which is a bit long for some of the experiments we want to do. So, it might be worth considering using DDR to update both DDSs at once. As the DDSs are identical, this doesn't introduce the kinds of corner cases that @dhslichter is worried about.

**Comment by gkasprow** [comment]
@hartytp @dhslichter So we could use double channel DDR SPI bus to control 4 DDS chips at once. This would consume 7 of 8 LVDS lines. The last one would be used for UPDATE signal.
We still can have a compromise here - we can route switches control both to TTL register and  to dedicated LVDS signals from second IDC.
So one who wants to have real time control of the switches can use second IDC but one who can live with slower control but need many DDS channels would use only single IDC and serial register.
The LVDS receivers we use have bus hold capability which means they ensure defined state when not connected to the source.
The outputs from LVDS receivers would be ORed with register outputs to make sure that both sources can control them..

**Comment by gkasprow** [comment]
@dhslichter Another issue is with AD9914. With second IDC connector, we will use 4 signals to control the output switches and remaining 4 to switch DDS profiles. But each DDS has 3 profile inputs. How would you like to connect them? One LVDS line per each DDS chip connected to one profile input or some fast serial register that handles all 4 chips * 3 inputs = 12 signals.

**Comment by gkasprow** [comment]
@jordens @hartytp @dhslichter 
What front panel LEDs should indicate?
- RF switch positions
- 9914 profile
- user defined functions (routed to TTL register)
- Overtermperature  / Power Good

**Comment by gkasprow** [comment]
@jordens @hartytp @dhslichter
I published still early version of [schematics](https://github.com/m-labs/sinara/blob/master/ARTIQ_ALTIUM/Kasli/3U/PCB_3U_DDS/3U_DDS.PDF) and semi-random component placement. So far only SDR SPI but with 2 channels is implemented.
I did complete power budget estimation. It's going to consume 12W in worst case for 9912 and almost twice more for 9914. I implemented hardware over-temperature protection.
I added 9914 option - it seems there is no compromise needed and we can fully implement two variants.
To finish schematics we have to decide about SPI mode (SDR vs DDR) and profile pins routing for 9914.

**Comment by hartytp** [comment]
@gkasprow Thanks for doing that @gkasprow! The schematics look really nice.

Should we consider using a different output amplifier? AFAICT, there's no need for a special low phase noise amp here (e.g. we're using the DDS's internal PLLs), so we could go for something which dissipates less power and is a bit cheaper. 

Not sure what P1dB people want for this design -- although, if it's used with the [RF PA](https://github.com/m-labs/sinara/wiki/RFPA), which has 40dB of gain, then even 0dBm is enough -- but, something like a AG203-63, which can run directly from a 3V3 rail might be a good choice. This is a little cheaper and would shave 1.6W off the design, which would make thermal management a bit easier.

**Comment by hartytp** [comment]
- [ ] If there is enough room on the front pane without cramming things, could we have a clock output as well. That way users can daisy-chain the clocks between modules. There are plenty of ADCLK948 outputs spare!

- [ ] Also, are we planning to float the SMA grounds by default? Or, should we just ground them, but leave the capacitor pads there so that they can be floated later if needed?

- [ ] Can we scrap the LTC6957-1? We've made some phase noise measurements with the ADCLK948 and a +10dBm sinusoid input at 100MHz and the performance is pretty similar. As this board isn't aiming for ultra-low noise, this part seems redundant IMO and removing it takes off another 200mW. I'd also recommend leaving the balun NC by default and replacing it with a 0R resistor in parallel with it.

**Comment by hartytp** [comment]
> What front panel LEDs should indicate?

How about a single bi-colour LED: green power/temp good; red over-temperature.

I don't see the point having LEDs to indicate switch/profile statuses.

**Comment by gkasprow** [comment]
@hartytp 
OK, I will add 0R resistors that connects SMA to GND. They can be removed when somebody needs isolation.
I have place for SMA clock output, I will add it.
OK, I will scrap LTC chip.
Two LEDs are easier to control so I'll leave as it is.

**Comment by hartytp** [comment]
@gkasprow Thinking more about thermal management: the LVDS interface takes up quite a lot of power in this design. Is there any way we can simplify it/reduce the power a bit without sacrificing performance?

- [ ] What default output filter do we want for this? I assume a 400MHz LP response using the MCL pads?

- [ ] Once the design has been finalised, please could you annotate the schematic with the RF output power and bandwidth?

**Comment by hartytp** [comment]
Another point for consideration: we can potentially shave another Watt from the power budget by replacing the ADCLK948 with an ADCLK846. This doesn't provide a mux, but personally I'd be happy with solder jumpers for that. My guess is that users won't switch clock sources often at all, so the solder jumper is fine.

**Comment by gkasprow** [comment]
@hartytp we can also use CDCLVD1208RHDT which consumes same amount of power but has input mux. Cost is also the same.

**Comment by hartytp** [comment]
@gkasprow From a quick look at the data sheets, that looks fine to me. Comments:

- Phase noise is better than the AD9912 with the PLL on
- Output frequency is 800MHz max, so we wouldn't be able to drive the AD9912 with a 1GHz external clock directly (not that we need to IMO).
- Users also need to be aware that although we using 3V3 LVPECL in most places, the clock output SMA from Urukul is LVDS (not that it will make a difference in most cases). Also, the phase noise will be slightly degraded compared with the ADCLK. This isn't a problem IMO, but users need to be told to take care when daisy-chaining an Urukul to a Sayma...

**Comment by gkasprow** [comment]
@hartytp  Generally, PECL devices have much better characteristics than LVDS ones due to much lower output resistance and sharper edges. But power consumption is real nightmare. This is the main reason why LVPECL is used mainly for clock distribution.
Is there any source of 1GHz clock which we would need to route to input of the clock buffer?
The output is not LVDS since it is single ended:) It is simply low amplitude AC coupled signal.

**Comment by hartytp** [comment]
@gkasprow Do you think we'll get decent performance using only half of the LVDS signal for the clock daisy-chaining (after including a bit of loss in cabling etc)? Or, do you think we'll need baluns on the clock input and output after all? If there is a doubt, then maybe we should add DNF pads for the baluns.

**Comment by dhslichter** [comment]
Can the clock daisy chaining circuit have a powerdown?  It would be nice to have a reasonable quality clock available if people want, but be able to DIP switch/jumper to disable (and thus get rid of the power dissipation too).

Waiting for @jordens to weigh in but to my mind, if you are driving AOMs, you'll need another power amp anyway so you might cut down on power dissipation by doing away with the output amp altogether, just using a reconstruction filter (user solderable) and the digital attenuator and switch.  This assumes sufficient gain in the power amps though (probably at least ~40 dB would be required).  

Most of the DDS chips will work with a sinusoidal input clock and a balun.  How about just using a passive splitter and requiring the user to supply all clocks externally?  This seems to me to be not so different in terms of the complexity of external wiring (you are still not sending any signals over a backplane, all on front panel), but it cuts down on power dissipation, will not degrade the clock quality, and can work for clocks up to 1 GHz.

**Comment by hartytp** [comment]
> Waiting for @jordens to weigh in but to my mind, if you are driving AOMs, you'll need another power amp anyway so you might cut down on power dissipation by doing away with the output amp altogether, just using a reconstruction filter (user solderable) and the digital attenuator and switch. This assumes sufficient gain in the power amps though (probably at least ~40 dB would be required).

Agreed, if one has a PA with decent gain then the output amp on Urukul is not required. I'd be happy scrapping it.

**Comment by hartytp** [comment]
> Most of the DDS chips will work with a sinusoidal input clock and a balun. How about just using a passive splitter and requiring the user to supply all clocks externally? This seems to me to be not so different in terms of the complexity of external wiring (you are still not sending any signals over a backplane, all on front panel), but it cuts down on power dissipation, will not degrade the clock quality, and can work for clocks up to 1 GHz.

I'd definitely prefer to keep the clock buffer IC on this design.

1. The CMOS fanout IC will dissipate < 150mW (in contrast, the ADCLK948 would have been closer to 1W IIRC). IMO, that's low enough not to worry about.

2. With passive 4-way splitting of the clock, the input clock has to have a reasonable amplitude to avoid seriously degrading the DDS noise floor. Assuming one has a few of Urukuls then one would need a beefy fan-out IC somewhere anyway to get the clock signal level high enough for passive splitting. In that case, it makes most sense to put the fan-out on Urukul itself rather than having to have it in a separate box.

3. Having a passive splitter also prevents daisy changing of Urukuls as the signal level will quickly get too low. My hope is that we can use the recovered clock from Kasli DRTIO as our 100MHz source for all our AOM DDSs. For that, we need a single clock output SMA on Kasli and the ability to daisy-chain it to the Urukuls.

4. The clock quality from the LVDS can still be very high quality: better than the DDS + PLL noise floor; better than the recovered clock from Kasli; and, plenty good enough for cooling/optical pumping/state preparation/etc that Urukul will be used for. Users just need to know not to try to use it to supply an ultra-low noise clock for Sayma.

5. True, the LVDS IC doesn't go up to 1GHz. But, do we really have a use-case for that? Otherwise let's not worry about it. The DDS + PLL noise floor is still really quite good and will be fine for any application I can imagine using Urukul for. NB 1GHz clocks also weren't supported by Greg's original design as the LTC IC only goes up to 300MHz IIRC.

> Can the clock daisy chaining circuit have a powerdown? It would be nice to have a reasonable quality clock available if people want, but be able to DIP switch/jumper to disable (and thus get rid of the power dissipation too).

Assuming one can just leave the SMA un-terimated when not in use, the clock daisy chaining circuit doesn't dissipate any power if it's not used. Thus, there's nothing to do here. Even if one does terminate it, the extra power consumption of a single LVDS line is pretty negligible for this design. I'm not sure that's the right place to try to optimise.

**Comment by jordens** [comment]
I'll catch up on the remaining pile of comments later. Thanks to all for participating in this!

AFAICT we are heading for two board version here. The DDS requirements for Oxford and optIclock are already different. And the IDC connector usage is as well. I have updated the current idea as far as I could on the wiki:

https://github.com/m-labs/sinara/wiki/Urukul#idc-usage

Let's try to discuss the individual topic areas in individual issues, like @sbourdeauducq did with #193

**Comment by jordens** [comment]
Let's retire this. We have:

#193 
#195 
#196 
#197 
#198 
#199 
#200 
#201

**Comment by hartytp** [comment]
@gkasprow What's the timeline for finishing the designs for Urukul & Kasli and ordering prototype hardware?

**Comment by gkasprow** [comment]
So far they are not funded yet, I mean the projects officially don't exist
at WUT. And we are focused on other hardware.
I've just published updated Kasli schematics which seem to be complete.
Once we accept them, can begin with routing.
